crates/edr_solidity/src/contracts_identifier.rs

```diff
@@ -0,0 +1,703 @@
+mod radix_tree;
+
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+};
+
+use radix_tree::RadixTree;
+use revm::interpreter::opcode;
+use revm_primitives::Bytes;
+
+use self::radix_tree::RadixNode;
+use crate::opcodes::get_opcode_length;
+
+#[derive(Debug, PartialEq)]
+pub enum BytecodeType {
+    Runtime,
+    Deployment,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ImmutableReference {
+    offset: usize,
+    length: usize,
+}
+
+// bytecode and trace stubs
+#[derive(Debug, PartialEq)]
+pub struct Bytecode {
+    normalized_code: Bytes,
+    bytecode_type: BytecodeType,
+    library_offsets: Vec<usize>,
+    immutable_references: Vec<ImmutableReference>,
+}
+
+impl Bytecode {
+    pub fn is_deployment(&self) -> bool {
+        self.bytecode_type == BytecodeType::Deployment
+    }
+}
+
+pub struct CreateMessageTrace {
+    code: Bytes,
+}
+
+pub struct CallMessageTrace {
+    code: Bytes,
+}
+
+pub enum EvmMessageTrace {
+    Create(CreateMessageTrace),
+    Call(CallMessageTrace),
+}
+
+impl EvmMessageTrace {
+    fn code(&self) -> &Bytes {
+        match self {
+            EvmMessageTrace::Create(create_message_trace) => &create_message_trace.code,
+            EvmMessageTrace::Call(call_message_trace) => &call_message_trace.code,
+        }
+    }
+}
+
+// TODO add cache
+#[derive(Default)]
+pub struct ContractsIdentifier<'a> {
+    tree: RadixTree,
+    bytecodes: HashMap<u64, &'a Bytecode>,
```

Comment:
I'm wondering if the `u64` space is large enough to avoid collisions? 

A [news article](https://cryptopotato.com/over-44-million-contracts-deployed-to-ethereum-since-genesis-research/) from 2022 indicates that 44 million smart contracts have been deployed on mainnet, so I think an estimate of 100m smart contracts existing across all chains is a reasonable estimate. With ~100m smart contracts we get roughly a 1 in 2000 probability for a collision using a 64-bit hash space, which is I think too high (see [here](https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation) for formula). 

If a collision can have security implications, I'd increase the hash space to 192-bit, as that'd result in a 2^-129 probability of collision for 2^32 smart contracts. Otherwise, I'd increase the hash space to 128-bit, as that'd result in a 2^-65 probability of collision which seems acceptable as it's comparable to correctly guessing the first time an 11-character alphanumeric password.

For the hash function I'd use [blake3](https://github.com/BLAKE3-team/BLAKE3) in both cases as it's a cryptographic hash function that's faster than md5.

Reply:
On a second reading of the code, I'm wondering if we could use the `Bytes` as keys directly? Then we could let the `HashMap` resolve collisions.

Reply:
Thanks, using the code as the key seems to work fine. Done in b2f692b193d14af6fff6683751888b6152a47769